azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/async_internal_client.py

```diff
@@ -0,0 +1,95 @@
+    """
+    logger.info("async connecting to transport")
+
+    # we need the event loop so we can run our code in side it.
+    loop = asyncio.get_event_loop()
+
+    # the future is how we communicate the result from the callback task back into the "main" task
+    future = asyncio.Future()
+
+    # This is a sync callback that is called by the transport.  We don't know what thread it's
+    # going to be called in.
+    def callback():
+        logger.info("async connect finished")
+
+        # We use the asyncio event loop from our outer scope because that's the event loop for the
+        # thread that started this operation.  This is important because this callback is called
+        # in the context of some transport thread, and we can almost guarantee that this thread
+        # does not have an asyncio event loop.  This line basically says "call future.set_result(None),
+        # as soon as you can, in the context of the (outer scope) asyncio event loop."
+        #
+        # Interestingly enough, even though future.set_result() is not an awaitable function, you
+        # still need to call it from inside the context of an asyncio event loop.  Otherwise it
+        # won't propertly signal any awaiters.  Don't get tricked by this and waste as much time
+        # as I did :)
+
+        loop.call_soon_threadsafe(future.set_result, None)
+
+    # call self._transport.connect(callback) in a threadpool thread managed by asyncio.
+    # This thread pool is called a ThreadPoolExecutor, which is a subclass of Executor. This way,
+    # the calling thread doesn't get blocked by our calls down into the transport.  In asyncio
+    # parlance, an Executor is something that lets you await blocking code.  ThreadPoolExecutor
+    # farms out the code to (CPUs * 5) background threads (by default).  There is also
+    # ProcessPoolExecutor which can farm the code out to background processes.  The first
+    # None parameter means "use the default executor."
+
+    await loop.run_in_executor(None, self._transport.connect, callback)
+
+    # finally, await the callback's call to future.set_result
+    await future
+
+
+async def _disconnect_async(self):
+    logger.info("async disconnecting from transport")
+    loop = asyncio.get_event_loop()
+    future = asyncio.Future()
+
+    def callback():
+        logger.info("async disconnect finished")
+        loop.call_soon_threadsafe(future.set_result, None)
+
+    await loop.run_in_executor(None, self._transport.disconnect, callback)
+    await future
+
+
+async def _send_event_async(self, event):
+    logger.info("async sending event")
+    loop = asyncio.get_event_loop()
+    future = asyncio.Future()
+
+    def callback():
+        logger.info("async sending finished")
+        loop.call_soon_threadsafe(future.set_result, None)
+
+    await loop.run_in_executor(None, self._transport.send_event, event, callback)
+    await future
```

Comment:
i'm going to make the same comment i always do.. this looks like a pattern ^^
maybe a bit early to refactor but something to track as APIs multiply.